File: scripts/check-nav-v2-consultation-server-adapter.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  consultationClientRequestId,
  consultationConversionToWizardDraft,
  consultationDecisionPresentation,
  consultationDecisionRpcPreview,
  consultationServerPayloadPreview,
  minimizeConsultationDetailResponse,
  minimizeConsultationQueueResponse
} from '../assets/js/nav-v2/consultation-server-adapter-v2.js';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const fixture = JSON.parse(fs.readFileSync(
  path.join(ROOT, 'fixtures/nav-v2-consultation-server-adapter-scenarios.json'),
  'utf8'
));

assert.equal(fixture.schema_version, 2);
assert.equal(fixture.synthetic_only, true);

for (const scenario of fixture.payload_cases) {
  const result = consultationServerPayloadPreview(scenario.input, {
    client_request_id: scenario.input.client_request_id
  });
  assert.equal(result.server_ready, true, `${scenario.id}: payload must be server-ready`);
  assert.ok(result.payload, `${scenario.id}: payload is required`);
  assert.equal(result.rpc_preview.name, 'nav_v2_create_consultation');
  assert.deepEqual(result.rpc_preview.args, { p_payload: result.payload });
  assert.equal(result.payload.client_request_id, scenario.input.client_request_id.toLowerCase());
  for (const [key, expected] of Object.entries(scenario.expected)) {
    if (key === 'question_contains') {
      for (const marker of expected) {
        assert.ok(result.payload.question.includes(marker), `${scenario.id}: question missing ${marker}`);
      }
      continue;
    }
    if (key === 'warning_contains') {
      assert.ok(result.adapter_warnings.some((warning) => warning.includes(expected)), `${scenario.id}: warning missing ${expected}`);
      continue;
    }
    if (key === 'forbidden_payload_keys') {
      for (const forbidden of expected) {
        assert.equal(Object.prototype.hasOwnProperty.call(result.payload, forbidden), false, `${scenario.id}: payload exposes ${forbidden}`);
      }
      continue;
    }
    assert.deepEqual(result.payload[key], expected, `${scenario.id}: ${key} mismatch`);
  }
  assert.equal(result.persistence.deal_created, false, `${scenario.id}: must not create deal`);
  assert.equal(result.persistence.backlog_created, false, `${scenario.id}: must not create backlog`);
  assert.equal(result.persistence.document_url_persisted, false, `${scenario.id}: URL must not persist`);
  assert.equal(result.persistence.idempotency_key_present, true, `${scenario.id}: idempotency key missing`);
}

const stableInput = fixture.payload_cases[0].input;
const firstPreview = consultationServerPayloadPreview(stableInput);
const repeatPreview = consultationServerPayloadPreview(stableInput);
assert.deepEqual(firstPreview.payload, repeatPreview.payload, 'same client request ID must create the same payload');

for (const scenario of fixture.idempotency_cases) {
  const normalized = consultationClientRequestId(scenario.value);
  assert.equal(Boolean(normalized), scenario.valid, `${scenario.id}: UUID validity mismatch`);
  if (scenario.normalized) assert.equal(normalized, scenario.normalized);
}
const missingId = consultationServerPayloadPreview({ ...stableInput, client_request_id: '' });
assert.equal(missingId.server_ready, false);
assert.ok(missingId.errors.some((message) => message.includes('client_request_id')));
const malformedId = consultationServerPayloadPreview({ ...stableInput, client_request_id: 'bad-id' });
assert.equal(malformedId.server_ready, false);
assert.equal(malformedId.rpc_preview, null);

const queue = minimizeConsultationQueueResponse(fixture.queue_case.input);
const queueText = JSON.stringify(queue);
for (const forbidden of fixture.queue_case.forbidden_after_minimization) {
  assert.equal(queueText.includes(`"${forbidden}"`), false, `queue contains forbidden key ${forbidden}`);
}
assert.equal(queue.items.length, 1);
assert.equal(queue.items[0].reference, 'Консультация 00000000');
assert.equal(queue.items[0].actionable_for_lawyer, true);

const detail = minimizeConsultationDetailResponse(fixture.detail_case.input);
assert.equal(detail.profile.email, undefined);
assert.equal(detail.consultation.seller_phone, undefined);
assert.equal(detail.permissions.admin, undefined);
assert.equal(detail.conversion_draft.client_name, undefined);
assert.equal(detail.conversion_draft.creates_deal, false);
assert.equal(detail.conversion_draft.creates_backlog, false);
assert.equal(detail.messages[0].author_email, undefined);
assert.equal(detail.messages[0].body, 'Нужна полная подготовка к задатку.');

for (const scenario of fixture.decision_cases) {
  const result = consultationDecisionRpcPreview(scenario.input);
  assert.equal(result.ok, scenario.valid, `${scenario.id}: decision validity mismatch`);
  assert.equal(result.transport_enabled, false);
  if (!scenario.valid) {
    assert.equal(result.rpc_preview, null);
    continue;
  }
  assert.equal(result.rpc_preview.name, 'nav_v2_decide_consultation');
  assert.deepEqual(result.rpc_preview.args, {
    p_consultation_id: scenario.input.consultation_id,
    p_decision: scenario.input.decision,
    p_body: scenario.input.body,
    p_conversion_mode: scenario.input.conversion_mode
  });
  assert.equal(result.presentation.next_status, scenario.next_status);
  assert.equal(
    result.presentation.requires_conversion_mode,
    scenario.input.decision === 'convert_to_preparation'
  );
}

const answerPresentation = consultationDecisionPresentation('answer');
assert.equal(answerPresentation.requires_conversion_mode, false);
const conversionPresentation = consultationDecisionPresentation('convert_to_preparation');
assert.deepEqual(conversionPresentation.conversion_modes, ['deposit', 'deal']);
assert.equal(consultationDecisionPresentation('bad'), null);

for (const scenario of fixture.conversion_cases) {
  const wizard = consultationConversionToWizardDraft(scenario.input);
  if (scenario.expected === null) {
    assert.equal(wizard, null, `${scenario.id}: unsafe conversion must be rejected`);
    continue;
  }
  for (const [key, expected] of Object.entries(scenario.expected)) {
    assert.deepEqual(wizard[key], expected, `${scenario.id}: ${key} mismatch`);
  }
  assert.equal(wizard.consultationId, scenario.input.consultation_id);
  assert.equal(Object.prototype.hasOwnProperty.call(wizard, 'client_name'), false);
  assert.equal(Object.prototype.hasOwnProperty.call(wizard, 'documents_url'), false);
}

console.log('Navigator v2 consultation server adapter regression passed: idempotent create preview, exact four-argument decisions, DTO minimization and safe conversion');

File: scripts/check_nav_v2_consultation_server_adapter.py
from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
ADAPTER = ROOT / "assets/js/nav-v2/consultation-server-adapter-v2.js"
UI = ROOT / "assets/js/nav-v2/consultation-v2.js"
FIXTURES = ROOT / "fixtures/nav-v2-consultation-server-adapter-scenarios.json"
SEMANTIC = ROOT / "scripts/check-nav-v2-consultation-server-adapter.mjs"
SERVER_CONTRACT = ROOT / "config/nav-v2-consultation-lifecycle-contract.json"
SERVER_PROTOTYPE = ROOT / "supabase/prototypes/nav_v2_consultation_lifecycle.sql"
DOC = ROOT / "docs/NAV_V2_CONSULTATION_SERVER_ADAPTER_2026-07-16.md"
WORKFLOW = ROOT / ".github/workflows/nav-v2-consultation-server-adapter.yml"


def require(text: str, markers: tuple[str, ...], label: str, errors: list[str]) -> None:
    for marker in markers:
        if marker not in text:
            errors.append(f"{label}: missing {marker!r}")


def main() -> int:
    errors: list[str] = []
    for path in (ADAPTER, UI, FIXTURES, SEMANTIC, SERVER_CONTRACT, SERVER_PROTOTYPE, DOC, WORKFLOW):
        if not path.exists():
            errors.append(f"missing {path.relative_to(ROOT)}")
    if errors:
        print("\n".join(errors))
        return 1

    adapter = ADAPTER.read_text(encoding="utf-8")
    ui = UI.read_text(encoding="utf-8")
    fixtures = json.loads(FIXTURES.read_text(encoding="utf-8"))
    contract = json.loads(SERVER_CONTRACT.read_text(encoding="utf-8"))
    prototype = SERVER_PROTOTYPE.read_text(encoding="utf-8")
    semantic = SEMANTIC.read_text(encoding="utf-8")
    doc = DOC.read_text(encoding="utf-8")
    workflow = WORKFLOW.read_text(encoding="utf-8")

    if contract.get("status") != "repository_only_prototype" or contract.get("production_applied") is not False:
        errors.append("server consultation contract must remain repository-only and non-production")
    if "-- REPOSITORY-ONLY PROTOTYPE." not in prototype:
        errors.append("server consultation SQL must remain a prototype")

    require(adapter, (
        "export function consultationServerPayloadPreview",
        "export function minimizeConsultationQueueItem",
        "export function minimizeConsultationQueueResponse",
        "export function minimizeConsultationDetailResponse",
        "export function consultationDecisionPresentation",
        "export function consultationConversionToWizardDraft",
        "request_type: REQUEST_TYPE_BY_STAGE[value.stage] || 'legal_answer'",
        "representation_model: REPRESENTATION_MAP[value.side] || 'unknown'",
        "has_external_documents: Boolean(value.documents_url)",
        "document_url_persisted: false",
        "deal_created: false",
        "backlog_created: false",
        "known_facts_preserved_in_question",
    ), ADAPTER.name, errors)

    for forbidden in (
        "document.", "window.", "localStorage", "sessionStorage", "rpc(", "fetch(", ".from(",
        "nav_v2_create_consultation", "nav_v2_decide_consultation", "nav_v2_close_consultation"
    ):
        if forbidden in adapter:
            errors.append(f"consultation adapter must remain pure and transport-free: {forbidden}")

    payload_start = adapter.find("const payload = {")
    payload_end = adapter.find("\n  };", payload_start)
    payload_block = adapter[payload_start:payload_end] if payload_start >= 0 and payload_end >= 0 else ""
    if not payload_block:
        errors.append("could not locate consultation payload allowlist")
    for forbidden_key in ("documents_url:", "document_source_url:", "known_facts:", "client_name:", "phone:", "email:"):
        if forbidden_key in payload_block:
            errors.append(f"future create payload contains forbidden key: {forbidden_key}")

    queue_keys = set(contract.get("queue_item_keys") or [])
    for key in queue_keys:
        if f"'{key}'" not in adapter:
            errors.append(f"adapter queue allowlist missing server key: {key}")
    for key in contract.get("queue_forbidden_keys") or []:
        if key in {"question", "body", "safe_reference", "document_source_url", "client_name", "phone", "email"}:
            if f"'{key}'" in adapter.split("const QUEUE_ITEM_KEYS", 1)[1].split("]);", 1)[0]:
                errors.append(f"adapter queue allowlist exposes forbidden key: {key}")

    require(ui, (
        "consultationServerPayloadPreview",
        "consultationServerReadiness",
        "renderServerReadiness",
        "Будущий серверный payload готов",
        "Сохранится только признак наличия документов, не сама ссылка",
        "Данные никуда не отправляются",
    ), UI.name, errors)
    for forbidden in (
        "rpc('nav_v2_create_consultation'", 'rpc("nav_v2_create_consultation"',
        "nav_v2_decide_consultation", "nav_v2_add_consultation_clarification",
        "nav_v2_close_consultation", ".from('nav_consultations_v2'", '.from("nav_consultations_v2"'
    ):
        if forbidden in ui:
            errors.append(f"consultation UI must not call undeployed server surface: {forbidden}")

    if fixtures.get("synthetic_only") is not True:
        errors.append("adapter fixtures must remain synthetic-only")
    if len(fixtures.get("payload_cases") or []) < 6:
        errors.append("adapter payload matrix must contain at least six cases")
    payload_ids = {case.get("id") for case in fixtures.get("payload_cases") or []}
    for required in (
        "basic_question", "deposit_partner", "mortgage_matcap", "document_url_stripped",
        "minor_codes_are_lossless_in_text", "spouse_is_preserved_in_text"
    ):
        if required not in payload_ids:
            errors.append(f"missing adapter scenario {required}")

    require(semantic, (
        "consultationServerPayloadPreview",
        "minimizeConsultationQueueResponse",
        "minimizeConsultationDetailResponse",
        "consultationDecisionPresentation",
        "consultationConversionToWizardDraft",
        "document_url_persisted",
        "Navigator v2 consultation server adapter regression passed",
    ), SEMANTIC.name, errors)

    require(doc, (
        "repository-only consumer contract",
        "URL не входит в server payload",
        "известные факты",
        "укрупнённую категорию",
        "queue DTO",
        "detail DTO",
        "не вызывает RPC",
        "Production gate",
        "Rollback",
    ), DOC.name, errors)

    require(workflow, (
        "python3 scripts/check_nav_v2_consultation_server_adapter.py",
        "node scripts/check-nav-v2-consultation-server-adapter.mjs",
        "node --check assets/js/nav-v2/consultation-server-adapter-v2.js",
        "node --check assets/js/nav-v2/consultation-v2.js",
        "nav-v2-consultation-server-adapter",
    ), WORKFLOW.name, errors)

    if errors:
        print("Navigator v2 consultation server adapter errors:")
        for error in errors:
            print(f"- {error}")
        return 1

    print(
        "Navigator v2 consultation server adapter passed: lossless question text, URL stripping, "
        "DTO minimization, decision copy, explicit conversion and no undeployed RPC calls"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
